### ISSUE TITLE
Support for custom datatypes as parse errors

### ISSUE THREAD

**Issue by wdanilo** [mrkkrp/megaparsec#97]
Hello!
First of all congratulations! I've been watching the state of the art of parsing in Haskell for long time and I was hoping trifecta will go the way megaparsec went. In fact I was planning to create my fork of Parsec to support the "recovery" feature and something that is yet missing from megaprasec - namely the ability to raise custom datatype-based errors, not limiting ourselves to strings only. Do you plan to add support for it in the future?

**Comment by mrkkrp** [comment]
Thanks! It's an interesting idea, do you have some specific design of how this can be added to the library?

**Comment by wdanilo** [comment]
It should be just added as a type parameter denoting what datatype I want to use when reporting errors. This datatype should implement some kind of a standard parser error type class that will handle all possible errors that can happen internally, like unexpected token or whatever. This way you can just `runParser` and get `Either MyErrorType Ok`. Does it clarify it to you?

**Comment by mrkkrp** [comment]
I see. There are several ways to add this functionality, so I just wanted to hear about the option that you have in mind. I do think it's a good feature and it will probably be in upcoming Megaparsec 5.0.0.

**Comment by wdanilo** [comment]
Glad to hear that! If you will have any specific questions regarding implementation / ideas - contact me whenever you'll need to! :)

**Comment by mrkkrp** [comment]
A couple of things in your approach I don't like. First, AFAICT, this will add one more type variable to `ParsecT` type, and thus will make signatures more verbose — I would like to avoid that. Another thing is that returned type should somehow represent “normal” exceptions (unexpected, expected stuff) and in addition to that we would like to pass additional, non-string info through parse errors. Type classes can solve this, but implementer of new error type will need to do work that is already done in `Text.Megaparsec.Error` in order to get get it right. Given such API I would personally defined something like:

```haskell
data CustomError = CustomError ParseError MyData
```

And then I would make it instance of type class that implements things related to parse errors. I would delegate most of work to internal `ParseError` where all things are already handled well out-of-the-box.

Another thing is that to use this feature, we need to essentially allow failing with already pre-constructed errors (additional work, additional place to make mistakes, for example there is nothing to protect you from giving it wrong position). This may be the price for additional power, but I like current approach where library takes care of error positions and you only tell it which messages to include.

----

Alternative approach I would like to propose is to make `[Message]` component of `ParseError` a polymorphic list (for now it's list of `String`s). Things that are `Typeable` and `ErrorComponent` (say it's a class that determines how to print its instances in error messages) would go in.

Pros:

* `ParsecT` signature remains the same

* API remains the same, we just change signature of already existing primitive `failure` to something like `failure :: (Typeable e, ErrorComponent e) => [e] -> m a`. Existing code still works (given we provide some instances of `ErrorComponent` out-of-box).

* When you need to extract things, pattern match and `cast` and you will get your data (if it's provided, or `Nothing` when error is generated by library's machinery).

Cons:

* Type of data is not passed around on type level, in signatures. You need to know it to get it. This is not very pretty, but this is the trade-off.

----

Another idea is to include additional data into `ParseError` as type variable, so by default it will be `ParseError Void`, but you can use something different. This is better than the idea with type classes, IMO.

----

Anyone is invited to propose ideas or express opininos. I will add “discussion” label to this.

**Comment by wdanilo** [comment]
@mrkkrp I don't think adding a new type parameter is a bad idea. You can provide a type alias where this type parameter would be set to a standard one. All other solutions are worse than that, especially your last one, which is not type safe and uses some kind of existential types, which always gives worse performance.

Anyway, the best solution I can think of is something like that: we should have a type class that is used for "stadnard parser errors", something like:

```
class IsParserError m e where
    formatError :: StandardParserError -> m e
```

where `StandardPArserError` is a data type which is used by `MegaParser` internally to indicate all `MegaParser` - related errors. This way I can very simple implement current behavior, by for example:

```
instance Monad m => IsParserError m String where
    formatError = return ∘ show
```

I can also use any custom datatype, which handles the standard error in any way I want.
You can also create some aliases like:
```newtype ParsecTE e s m a = ...
type ParsecT = ParsecTE String```

**Comment by mrkkrp** [comment]
You basically re-iterate your original proposition here. I agree that existential things are of course slower (GHC must keeps things polymorphic at run time and pass dictionaries around, etc.), and they are not as type safe as eplicit typing. But why do you think that type classes for error-like things is better than inclusion of optional custom data type into `ParseError`, like `ParserError e`? Then we could simply add this `e` as argument to `failure`.

**Comment by wdanilo** [comment]
Yes, I just wanted to emphasize some of the most important points here. I'm thinking about additional type variable to `ParseError` - maybe it is even better solution than the proposed by me - because you've got more explicit types flowing around. If it would give me the same abilities ("throwing" my custom data-types based errors while parsing), I think it could be better solution. I'm not sure yet, because it really depends on how the types will be designed and which of these solutions would give simpler architecture, but I'm biased towards your proposition.

**Comment by mrkkrp** [comment]
OK, I'm going to play with this in #99 (among other things). I'll ping you when it's in more-or-less working state.

**Comment by wdanilo** [comment]
That would be great! Thank you :)

**Comment by mrkkrp** [comment]
Interestingly, custom data needs to be a monoid, because when parse errors are merged from different branches of parsing (this is monoidal append in Megaparsec), and position of parse error is the same we actually merge collections of messages. If we have something else in parse error, there is nothing reasonable we can do but to merge it as well (`mappend`).

From this it follows that we also should use `()` as default data type, not `Void`, because `()` is a monoid, while `Void` is not.

**Comment by wdanilo** [comment]
This sound reasonable and `()` is more Haskell-way to handle it, so yes, that sounds good so far.

**Comment by mrkkrp** [comment]
I would say `()` is just more often used. `Void` means total absence of something and it's used as such, in `conduit` for example to represent absence of source, etc.

**Comment by wdanilo** [comment]
Correct, saying "haskell-way" I ment "used by the majority of libraries in such purpose and even introduced by some books as "lack of data". If you consider standard monadic operators that return `m ()` you can assume that `()` is just treated as "the haskell way to express lack of data`. Personally I think it would be better to use some more explicit data than `empty tuple`, but it's just impossible to change it in every existing library. On the other hand I don't see (yet) nothing wrong in implementing `Monoid` instance for `Void` - it matches all the `Monoid` rules and seems pretty resonable.

**Comment by mrkkrp** [comment]
In `pFail` (`fail`) and `pZero` (`mzero` and `empty`), when signalling error, there is no other choice than bottom (`undefined`) to put into `ParseError`. This however will likely blow in runtime because of `mappend`ing. I'm in doubts about the whole idea now, but I want to finish this feature to see what I'll get in the end. I will continue replacing `a` with `Maybe a` in `ParseError`.

**Comment by mrkkrp** [comment]
Oh, but wait, we can put `mempty` there. I need a break :-D

**Comment by mrkkrp** [comment]
BTW, while this feature is certainly nice to have, it changes quite a lot in the library. Can you give “real-world” example where it would be super-useful?

**Comment by wdanilo** [comment]
Of course. I'm writing a compiler now. Part of a compiler (as everybody knows) is parsing. The compiler consist of passes communicating with each other with data. The parser is run many times, especially in the interactive mode and it passes the results to following passes. These passes take action based on the output and the action differs based on the error type. If there was a syntax error - we should disable this "part of the program from further evaluation (we've got here funny partial-program-evaluation-engine). If there was a unknown method error (checked while parsing, because we support multi-arguments functions, like `if then else` - you take another action. On fact there is a ton of such use cases.

**Comment by mrkkrp** [comment]
Good, I think I just realized how to better represent errors in new version of Megaparsec. In fact, what we have right now is from Parsec's past, and it was not designed to support any sort of custom error messages in the first place. Now I should certainly reorganize `ParseError` to encourage properly typed errors that can describe what is exactly wrong (while preserving support for the default “unexpected A, expecting B” case). Custom error components in form of strings should be removed.

**Comment by wdanilo** [comment]
That's exactly how I imagine that too :)

**Comment by mrkkrp** [comment]
See progress in #99. There are lots of work to do (and I hope I will have more time to hack it soon), but it should be obvious in what direction Megaparsec 5 is moving (in particular see how `ParseError` is organized). What do you think?

**Comment by mrkkrp** [comment]
I took a composite approach in #99. `SourcePos` looks like this now:

```
data MessageItem t
  = Token t                  -- ^ Single token
  | TokenStream (NonEmpty t) -- ^ Non-empty stream of tokens
  | Label (NonEmpty Char)    -- ^ Label (cannot be empty)
  | EndOfInput               -- ^ End of input
  deriving (Show, Read, Eq, Ord, Data, Typeable)

data ParseError t e = ParseError
  { errorPos        :: NonEmpty SourcePos  -- ^ Stack of source positions
  , errorUnexpected :: Set (MessageItem t) -- ^ Unexpected items
  , errorExpected   :: Set (MessageItem t) -- ^ Expected items
  , errorData       :: Set e               -- ^ Associated data, if any
  } deriving (Show, Read, Eq, Typeable)
```

`t` is type of token and `e` is type of custom information. I use `Set` instead of initial idea with monoidal `e`, I think it's better because I can imagine `e` is rarely something monoidal unless you make it some sort of collection anyway.

Default data type that is used as `e` is `Dec`:

```
-- | “Default error component”. This in our instance of 'ErrorComponent'
-- provided out-of-box.
--
-- @since 5.0.0

data Dec
  = DecFail String         -- ^ 'fail' has been used in parser monad
  | DecIndentation Ordering Pos Pos -- ^ Incorrect indentation error
  deriving (Show, Read, Eq, Ord, Data, Typeable)
```

Type class looks like this:

```
class Ord e => ErrorComponent e where

  -- | Represent message passed to 'fail' in parser monad.

  representFail :: String -> e

  -- | Represent information about incorrect indentation.

  representIndentation
    :: Ordering -- ^ Desired ordering between reference level and actual level
    -> Pos             -- ^ Reference indentation level
    -> Pos             -- ^ Actual indentation level
    -> e
```

This allowed me to improve error messages for indentation-sensitive parsing significantly (something I wanted to do anyway, #90). Rendering of parse errors is a separate thing on its own now, it's also more flexible.

Stuff in #99 finally passes the tests. There is more work to do, but I think the whole approach won't change much. So, what do you say now, are we close to closing the issue as solved?

**Comment by mrkkrp** [comment]
I consider this resolved via #99 now.

**Comment by mrkkrp** [comment]
@wdanilo, Have you lost interest in this feature? It would be great to get some feedback before public release, it will be too late then.
